src/daemon/ca/aspa.rs

```diff
@@ -0,0 +1,280 @@
+
+/// This type contains the ASPA definitions for a CA. Generally speaking
+/// the [`AspaCustomer`] ASN will be held in a single [`ResourceClass`] only,
+/// but at least in theory the CA could issue ASPA objects in each RC that
+/// holds the ASN.
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
+pub struct AspaDefinitions {
+    attestations: HashMap<AspaCustomer, AspaDefinition>,
+}
+
+impl AspaDefinitions {
+    // Add or replace a new definition
+    pub fn add_or_replace(&mut self, aspa: AspaDefinition) {
+        let customer = aspa.customer();
+        self.attestations.insert(customer, aspa);
+    }
+
+    // Remove an existing definition (if it is present)
+    pub fn remove(&mut self, customer: AspaCustomer) {
+        self.attestations.remove(&customer);
+    }
+
+    // Applies an update. This assumes that the update was verified beforehand.
+    pub fn apply_update(&mut self, customer: AspaCustomer, update: &AspaConfigurationUpdate) {
+        let current = self.attestations.get_mut(&customer).unwrap();
+        current.apply_update(update);
+    }
+
+    pub fn all(&self) -> impl Iterator<Item = &AspaDefinition> {
+        self.attestations.values()
+    }
+}
+
+/// # Set operations
+///
+impl AspaDefinitions {
+    pub fn get(&self, customer: AspaCustomer) -> Option<&AspaDefinition> {
+        self.attestations.get(&customer)
+    }
+
+    pub fn has(&self, customer: AspaCustomer) -> bool {
+        self.attestations.contains_key(&customer)
+    }
+
+    pub fn len(&self) -> usize {
+        self.attestations.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.attestations.is_empty()
+    }
+}
+
+impl Default for AspaDefinitions {
+    fn default() -> AspaDefinitions {
+        AspaDefinitions {
+            attestations: HashMap::new(),
+        }
+    }
+}
+
+//------------ AspaObjects -------------------------------------------------
+
+/// ASPA objects held by a resource class in a CA.
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
+pub struct AspaObjects(HashMap<AspaCustomer, AspaInfo>);
+
+impl AspaObjects {
+    pub fn make_aspa(
+        &self,
+        aspa_definition: AspaDefinition,
+        certified_key: &CertifiedKey,
+        issuance_timing: &IssuanceTimingConfig,
+        signer: &KrillSigner,
+    ) -> KrillResult<AspaInfo> {
+        let weeks = issuance_timing.timing_aspa_valid_weeks;
+        let aspa = make_aspa_object(aspa_definition.clone(), certified_key, weeks, signer)?;
+        Ok(AspaInfo::new_aspa(aspa_definition, aspa))
+    }
+
+    /// Issue new ASPA objects based on configuration, and remove
+    /// object for which the customer AS is no longer held.
+    ///
+    /// Note: we pass in *all* AspaDefinitions for the CA, not all
+    ///   definitions will be relevant for the RC (key) holding
+    ///   this AspaObjects.
+    pub fn update(
+        &self,
+        all_aspas: &AspaDefinitions,
```

Comment:
As neither `all_aspas` nor `relevant_aspa` are actually `Aspa` instances it would be less confusing to include `def` or `definition` in these variable names as they are instances of the `AspaDefinition` type.

Reply:
done